Libraries/LibCore/EventLoop.cpp

```diff
@@ -82,13 +83,65 @@ static NeverDestroyed<IDAllocator> s_id_allocator;
 static HashMap<int, NonnullOwnPtr<EventLoopTimer>>* s_timers;
 static HashTable<Notifier*>* s_notifiers;
 int EventLoop::s_wake_pipe_fds[2];
-HashMap<int, EventLoop::SignalHandlers> EventLoop::s_signal_handlers;
-int EventLoop::s_handling_signal = 0;
-int EventLoop::s_next_signal_id = 0;
-pid_t EventLoop::s_pid;
 static RefPtr<LocalServer> s_rpc_server;
 HashMap<int, RefPtr<RPCClient>> s_rpc_clients;
 
+class SignalHandlers : public RefCounted<SignalHandlers> {
+    AK_MAKE_NONCOPYABLE(SignalHandlers);
+    AK_MAKE_NONMOVABLE(SignalHandlers);
+
+public:
+    SignalHandlers(int signo, void (*handle_signal)(int));
+    ~SignalHandlers();
+
+    void dispatch();
+    int add(Function<void(int)>&& handler);
+    bool remove(int handler_id);
+
+    bool is_empty() const
+    {
+        if (m_calling_handlers) {
+            for (auto& handler : m_handlers_pending) {
+                if (handler.value)
+                    return false; // an add is pending
+            }
+        }
+        return m_handlers.is_empty();
+    }
+
+    bool have(int handler_id) const
+    {
+        if (m_calling_handlers) {
+            auto it = m_handlers_pending.find(handler_id);
+            if (it != m_handlers_pending.end()) {
+                if (!it->value)
+                    return false; // a deletion is pending
+            }
+        }
+        return m_handlers.contains(handler_id);
+    }
+
+    int m_signo;
+    void (*m_original_handler)(int); // TODO: can't use sighandler_t?
+    HashMap<int, Function<void(int)>> m_handlers;
+    HashMap<int, Function<void(int)>> m_handlers_pending;
+    bool m_calling_handlers { false };
+};
+
+struct SignalHandlersInfo {
+    HashMap<int, NonnullRefPtr<SignalHandlers>> signal_handlers;
+    int next_signal_id { 0 };
+};
+
+template<bool create_if_null = true>
+inline SignalHandlersInfo* signals_info()
+{
+    static SignalHandlersInfo* s_signals;
+    return AK::Singleton<SignalHandlersInfo>::get(s_signals);
```

Comment:
I'm curious, what benefit does creating an AK::Singleton object have over using a function-scoped static? I guess, it's storage is on the heap rather than the `.data` segment of the object? But all of the thread-safe only-init-once stuff is handled by __cxa_guard_acquire, __cxa_guard_release, and __cxa_guard_abort in the standard library, as shown in my failed attempt to remove libstdc++ as a dependency last year (dang ports!). https://github.com/SerenityOS/serenity/pull/2242/files#diff-adcd02a2365753c4ecc8545590bfe81a1a8b2285da29c595952951ca4ec06bf8R36

Reply:
It's essentially the same thing, except that you have the compiler generate the magic code rather than writing it explicitly. Did that get checked in?

Reply:
It got merged in but reverted because the gcc port needed a fully featured libstdc++. Right now any function scope statics call into the __cxa_guard_foo functions provided by libgcc_s.